### PR TITLE
AppleFrameworks: Use ExtraFrameworkDependency internally

### LINKED ISSUE #3207: AppleFrameworks: Actually check if the frameworks are available
(I see it is already a FIXME, just making it an issue for visibility)

https://github.com/mesonbuild/meson/blob/b8f02047bec9fd2d1a36db82df5fed14ef386cd6/mesonbuild/dependencies/platform.py#L23-L40
When finding dependencies which may be frameworks on macOS, I'd like to be able to check for it in pkg-config (in-case of a custom one made by the user, Homebrew, etc), and if it's not found and macOS then check if it exists as a Framework. Obviously as it stands, this will never fail on macOS.

### PR DESCRIPTION
**Pull request by januaryjp**
Also add 'include' option for extra Framework search paths. Fixes #3207.

### CODE REVIEW

File: mesonbuild/dependencies/platform.py [hunk 1]
```diff
@@ -17,24 +17,43 @@
...
+
+        for f in modules:
+            self.frameworks += [ExtraFrameworkDependency(
+                f,
+                True,
+                includes,
+                env,
+                None,
+                kwargs,
+                )]
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/platform.py]: __[Flake8]__

[E123] closing bracket does not match indentation of opening bracket's line

File: mesonbuild/dependencies/platform.py [hunk 2]
```diff
@@ -17,24 +17,43 @@
...
+
         if isinstance(modules, str):
             modules = [modules]
         if not modules:
             raise DependencyException("AppleFrameworks dependency requires at least one module.")
-        self.frameworks = modules
-        # FIXME: Use self.compiler to check if the frameworks are available
+
+        if isinstance(includes, str):
+            includes = [include]
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/platform.py]: __[Flake8]__

[F821] undefined name 'include'

**Comment by jon-turney** [comment]
Since PR#3214 has landed, AppleFramework is now documented in `docs/markdown/Dependencies.md`, so `include:` should be added there.

Does `include:` really make sense as the name for an option indicating extra search paths?

Also, it looks to me like `find_external_dependency` will look for dependencies as frameworks on osx, if they can't be found with pkg-config, so .... um... it beats me when this is supposed to be used...

File: mesonbuild/dependencies/base.py [hunk 3]
```diff
@@ -878,19 +878,19 @@ def __init__(self, name, required, path, env, lang, kwargs):
...
             mlog.log('Dependency', mlog.bold(name), 'found:', mlog.green('YES'),
                      os.path.join(self.path, self.name))
         else:
             mlog.log('Dependency', name, 'found:', mlog.red('NO'))
 
-    def detect(self, name, path):
+    def detect(self, name, paths):
         lname = name.lower()
-        if path is None:
+        if paths is None or len(paths) == 0:
```
Comment (jon-turney) [inline mesonbuild/dependencies/base.py]: both None and empty sequences are false, so this can just be `if paths:` ?

**Comment by eintw1ck** [comment]
@jon-turney The idea behind this is that some libraries under macOS have different names and are frameworks. For example, Vulkan is 'Vulkan' on everything else, except in macOS it's a .framework and it's called MoltenVK.framework. So if there were a way to specify a macOS framework fallback to dependency() then there's be no need to have 'appleframeworks'.  `include:` probably isn't the best name, but I thought that `extra_framework_paths:` would be too long.

**Comment by jpakkane** [comment]
One thing which we have been considering is to have multiple arguments to `dependency` just like for `find_program` which would mean "use the first one of these that can be found" so you could do:

    vk_dep = dependency('vulkan', 'MoltenVK')

This is trickier because of subproject fallbacks.

**Comment by eintw1ck** [comment]
This could work, though I still feel that you should be able to confine certain dependency fallbacks to a specific OS.

**Comment by jpakkane** [comment]
Could a test be added for this so we don't break this in the future?

**Comment by eintw1ck** [comment]
I thought you wanted to get `dependency()` dependency fallbacks first?

**Comment by januaryjp** [comment]
I don’t really care to be honest. For me: as it currently is it doesn’t work—I want it fixed. If it doesn’t fit your pretty ideology of how you want it done then fix it some other way

**Comment by nirbheek** [comment]
> I don’t really care to be honest. For me: as it currently is it doesn’t work—I want it fixed. If it doesn’t fit your pretty ideology of how you want it done then fix it some other way

Being abrasive is not how you convince people and get things done, and such tone is not welcome in this project. I understand that it can be frustrating to wait months before getting a reply, but that's just how a volunteer-driven project works.

Naming things is hard, and Jussi was merely trying to say that we should find a better name for it, perhaps something that makes it clear that it only applies to frameworks and also takes a list of paths, such as `framework_extra_paths:`

**Comment by januaryjp** [comment]
No longer needed.

**Comment by nirbheek** [comment]
> Not because it is fixed, but because I can't be bothered to waste my time on meson anymore

I'm not sure what happened here or what we did wrong, and I am sorry if we inadvertently offended you in some way.

This PR looked fine to me except for the name, and you're probably the best person to fix this right now since I have my hands full and I don't use macOS on a daily basis.

**Pull request closed without merging**